File: src/days/day6/1.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func readGroupsFromFile(path string) [][]string {
	b, _ := ioutil.ReadFile(path) // just pass the file name
	trimmed := strings.TrimRight(string(b), "\n")
	groupsStr := strings.Split(trimmed, "\n\n")
	groups := make([][]string, 0)
	for _, g := range groupsStr {
		people := strings.Split(g, "\n")
		groups = append(groups, people)
	}
	return groups
}

func getNumberOfQuestionEverAnswered(groups [][]string) int {
	count := 0
	for _, group := range groups {
		answered := make(map[rune]int, 0)
		for _, person := range group {
			for _, question := range person {
				answered[question] = 1
			}
		}
		count += len(answered)
	}
	return count
}

func getNumberOfQuestionsThatEveryoneAnswered(groups [][]string) int {
	count := 0
	for _, group := range groups {
		answered := make(map[rune]int, 0)
		numberOfPeople := len(group)
		for _, person := range group {
			for _, question := range person {
				answered[question] += 1
			}
		}
		for _, answeredTimes := range answered {
			if answeredTimes == numberOfPeople {
				count += 1
			}
		}
	}
	return count
}

func main() {
	groups := readGroupsFromFile("./in")

	// part 1
	everAnswered := getNumberOfQuestionEverAnswered(groups)
	fmt.Println(everAnswered)

	// part 2
	everyoneAnswered := getNumberOfQuestionsThatEveryoneAnswered(groups)
	fmt.Println(everyoneAnswered)
}

File: src/days/day6/1.py
from collections import defaultdict


def read_group_questions():
    f = open('in', 'r')
    groups = f.read().strip("\n").split("\n\n")
    return list(map(lambda x: x.split('\n'), groups))


def get_number_of_questions_ever_answered(groups):
    count = 0
    for group in groups:
        answered = {}
        for person in group:
            for question in person:
                answered[question] = True
        count += len(answered)
    return count


def get_number_of_questions_that_everyone_answered(groups):
    count = 0
    for group in groups:
        answered = defaultdict(int)
        people_count = len(group)
        for person in group:
            for question in person:
                answered[question] += 1
        for _, value in answered.items():
            if value == people_count:
                count += 1
    return count


def exec_part1(groups):
    return get_number_of_questions_ever_answered(groups)


def exec_part2(groups):
    return get_number_of_questions_that_everyone_answered(groups)


questions = read_group_questions()

# part 1
print(exec_part1(questions))

# part 2
print(exec_part2(questions))
